iceoryx_dust/vocabulary/include/iox/span.hpp

```diff
@@ -0,0 +1,451 @@
+/// @tparam T Type of the array
+/// @tparam N Size of the array
+/// @param An array
+/// @return Size of the array
+template <typename T, std::uint64_t N>
+constexpr std::uint64_t size(const T (&)[N]) noexcept;
+
+/// @brief Returns N
+/// @tparam T Type of the iox::UninitializedArray
+/// @tparam N Size of the iox::UninitializedArray
+/// @param An iox::UninitializedArray
+/// @return Size of the iox::UninitializedArray
+template <typename T, std::uint64_t N, template <typename, uint64_t> class Buffer>
+constexpr std::uint64_t size(const UninitializedArray<T, N, Buffer>&) noexcept;
+
+/// @brief Returns a pointer to the block of memory containing the elements of the range.
+/// @tparam C Type of the container
+/// @param c A container or view with a data() member function
+/// @return Returns c.data()
+template <typename Container>
+constexpr auto data(Container& container) -> decltype(container.data());
+
+/// @brief Returns a pointer to the block of memory containing the elements of the range.
+/// @tparam C Type of the container
+/// @param c A container or view with a data() member function
+/// @return Returns c.data()
+template <typename Container>
+constexpr auto data(const Container& container) -> decltype(container.data());
+
+/// @brief Returns a pointer to the block of memory containing the elements of the range.
+/// @tparam T Type of the array
+/// @tparam N Size of the array
+/// @param array An array of arbitrary type
+/// @return Returns array
+template <typename T, std::uint64_t N>
+// NOLINTNEXTLINE(cppcoreguidelines-avoid-c-arrays, hicpp-avoid-c-arrays)
+constexpr T* data(T (&array)[N]) noexcept;
+
+/// @brief Returns a pointer to the block of memory containing the elements of the range.
+/// @tparam T Type of the iox::UninitializedArray
+/// @tparam N Size of the iox::UninitializedArray
+/// @param array An iox::UninitializedArray of arbitrary type
+/// @return Returns iox::UninitializedArray
+template <typename T, std::uint64_t N, template <typename, uint64_t> class Buffer>
+constexpr T* data(UninitializedArray<T, N, Buffer>& uninit_array) noexcept;
+
+template <typename T, uint64_t Extent = DYNAMIC_EXTENT>
+class span;
+
+namespace detail
+{
+template <uint64_t I>
+using size_constant = std::integral_constant<uint64_t, I>;
+
+template <typename T>
+struct extent_impl : size_constant<DYNAMIC_EXTENT>
+{
+};
+
+template <typename T, uint64_t N>
+// NOLINTNEXTLINE(cppcoreguidelines-avoid-c-arrays, hicpp-avoid-c-arrays)
+struct extent_impl<T[N]> : size_constant<N>
+{
+};
+
+template <typename T, uint64_t N>
+struct extent_impl<iox::UninitializedArray<T, N>> : size_constant<N>
+{
+};
+
+template <typename T, uint64_t N>
+struct extent_impl<iox::span<T, N>> : size_constant<N>
+{
+};
+
+template <typename T>
+using extent_t = extent_impl<remove_cvref_t<T>>;
+
+template <typename T>
+struct is_span : std::false_type
+{
+};
+
+template <typename T, uint64_t Extent_t>
+struct is_span<span<T, Extent_t>> : std::true_type
```

Comment:
```suggestion
template <typename T, uint64_t Extent>
struct is_span<span<T, Extent>> : std::true_type
```

It is a little bit inconsistent since the other template parameter also do not have a `_t` suffix